### PR TITLE
[lib] nixosSystem: allow setting global/nixos/hm modules and update to latest HM

### PR DESCRIPTION
**Pull request by kalbasit**
The HM flake has been updated to allow setting special arguments as well as modules directly through the `home-manager` top-level option. This PR takes advantage of that and allows setting special arguments and modules globally, per NixOS and per HM configuration.

NOTE: This requires updated your soxincfg to reflect the change of modules -> globalModules.
NOTE: This was extracted from #29 and tested only in that branch.

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -40,9 +40,7 @@
 
         overlays = import ./overlays;
 
-        nixosModules = (import ./modules) // {
-          soxin = import ./modules/soxin.nix;
-        };
+        nixosModules.soxin = import ./modules/soxin.nix;
```
Comment (rissson) [inline flake.nix:43]: Does that mean we don't want to expose individual modules? For example, I would have liked to reuse the bluetooth module for a work project. I know I can import everything and just turn the bluetooth on, but that means having blind trust into the upstream project that nothing is set by default, and it also means that upgrading requires reviewing every single change to make sure of that, instead of only reviewing the changes to the list of modules (to make sure the bluetooth module hasn't added some other files) and to the bluetooth module.

File: lib/nixos-system.nix [hunk 2]
```diff
@@ -1,43 +1,93 @@
...
+  # NixOS specific extra arguments.
+, nixosSpecialArgs ? { }
+
+, ...
+} @ args:
+let
+  args' = removeAttrs args [
+    "globalModules"
+    "nixosModules"
+    "hmModules"
```
Comment (rissson) [inline lib/nixos-system.nix:30]: The ordering of those is not the same as above. (Sorry for nitpicking)

File: lib/nixos-system.nix [hunk 3]
```diff
@@ -1,43 +1,93 @@
...
+    "hmSpecialArgs"
+  ];
+in
+nixosSystem (recursiveUpdate args' {
   specialArgs = {
-    mode = "NixOS";
-    soxin = self;
-  } // globalSpecialArgs // nixosSpecialArgs;
+    # send home-manager down to the NixOS modules
+    inherit home-manager;
```
Comment (rissson) [inline lib/nixos-system.nix:40]: Why is this required? i.e. why pass it as a specialArg instead of `_module.args` as it's currently done?

File: lib/nixos-system.nix [hunk 4]
```diff
@@ -1,43 +1,93 @@
...
-    {
+  modules =
+    # include the global modules
+    globalModules
+    # include the NixOS modules
+    ++ nixosModules
+    # include all Soxin modules
+    ++ (builtins.attrValues self.nixosModules)
+    # include all home-manager modules
+    ++ (builtins.attrValues home-manager.nixosModules)
```
Comment (rissson) [inline lib/nixos-system.nix:61]: Let's maybe more specific in what we want to import. I know that we manage this project and we should check nothing is added here that we don't want, but in case home-manager adds an unwanted module, let's be sure we only include the home-manager one, instead of everything.

File: flake.nix (hunk 1 above)
Comment (rissson) [inline flake.nix:43]: We should also add `nixosModule = self.nixosModules.soxin;`

File: lib/nixos-system.nix [hunk 5]
```diff
@@ -1,43 +1,93 @@
...
-    })
-  ];
+
+      home-manager.extraSpecialArgs = {
+        # send home-manager down to the home-manager modules
+        inherit home-manager;
+
+        # the mode allows us to tell at what level we are within the modules.
+        mode = "home-manager";
+        # send soxin to all NixOS modules
```
Comment (rissson) [inline lib/nixos-system.nix:77]: I think you mean `home-manager` instead of `NixOS` here.

File: lib/nixos-system.nix [hunk 6]
```diff
@@ -1,43 +1,93 @@
...
-            soxin = self;
-          } // globalSpecialArgs // hmSpecialArgs;
-        });
-      };
-    })
-  ];
+
+      home-manager.extraSpecialArgs = {
+        # send home-manager down to the home-manager modules
+        inherit home-manager;
```
Comment (rissson) [inline lib/nixos-system.nix:73]: Same as above.

File: lib/nixos-system.nix [hunk 7]
```diff
@@ -1,43 +1,93 @@
...
+    mode = "NixOS";
 
-    self.nixosModules.soxin
+    # send soxin to all NixOS modules
+    soxin = self;
+  }
+  # include the global special arguments.
+  // globalSpecialArgs
+  # include the NixOS special arguments.
+  // nixosSpecialArgs;
```
Comment (rissson) [inline lib/nixos-system.nix:51]: The priority of the ordering here looks good to me. I wonder if we should use `recursiveUpdate` though.

File: lib/nixos-system.nix [hunk 8]
```diff
@@ -1,43 +1,93 @@
...
+
+        # the mode allows us to tell at what level we are within the modules.
+        mode = "home-manager";
+        # send soxin to all NixOS modules
+        soxin = self;
+      }
+      # include the global special arguments.
+      // globalSpecialArgs
+      # include the NixOS special arguments.
+      // hmSpecialArgs;
```
Comment (rissson) [inline lib/nixos-system.nix:83]: Same as above.

**Review by rissson — COMMENTED** [review]
Re-reading through my own review, it seems I have an occupational hazard, i.e. not trusting stuff that I use, but it's still worth asking those questions.

File: flake.nix (hunk 1 above)
Comment (kalbasit) [reply flake.nix:43]: I have not thought of that use case. Do you think they would individually fit the top-level nixosModules attribute or should it go in a sub-attribute?

Comment (kalbasit) [reply flake.nix:43]: Done.

File: lib/nixos-system.nix (hunk 2 above)
Comment (kalbasit) [reply lib/nixos-system.nix:30]: Done

File: lib/nixos-system.nix (hunk 3 above)
Comment (kalbasit) [reply lib/nixos-system.nix:40]: It works better. This solved one infinite recursion I was having while iterating on this.

File: lib/nixos-system.nix (hunk 7 above)
Comment (kalbasit) [reply lib/nixos-system.nix:51]: recursiveUpdate may be really slow if it ends up merging `pkgs` with `pkgs`. That's why I thought this should suffice for now.

File: lib/nixos-system.nix (hunk 4 above)
Comment (kalbasit) [reply lib/nixos-system.nix]: done

File: lib/nixos-system.nix (hunk 6 above)
Comment (kalbasit) [reply lib/nixos-system.nix:73]: Resolving this to the discussion centralized.

File: lib/nixos-system.nix (hunk 5 above)
Comment (kalbasit) [reply lib/nixos-system.nix:77]: done

File: lib/nixos-system.nix (hunk 8 above)
Comment (kalbasit) [reply lib/nixos-system.nix:83]: done

File: flake.nix (hunk 1 above)
Comment (rissson) [reply flake.nix:43]: I'm not sure flakes allow for modules to be nested in `nixosModules`.

File: lib/nixos-system.nix (hunk 7 above)
Comment (rissson) [reply lib/nixos-system.nix:51]: Indeed, I forgot that you would probably pass down some `pkgs` at some point.

File: flake.nix (hunk 1 above)
Comment (kalbasit) [reply flake.nix:43]: Let me add it back as it was then.

Comment (kalbasit) [reply flake.nix]: done

**Review by rissson — APPROVED** [review]
